From e3675873f366374a5539dd7d1bdfdc10a935f333 Mon Sep 17 00:00:00 2001
From: Takanari Hayama <taki@igel.co.jp>
Date: Mon, 29 Sep 2014 13:39:02 +0900
Subject: [PATCH 13/22] v4l2-renderer: check validity of DMABUF before
 composiiton.

OpenGL/ES clients may destroy buffers before they remove themselves
from the server. We need to check if the buffer is still valid before
trying to start composition.

This is not cheap, but we have no choice for now.
---
 src/v4l2-renderer.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/v4l2-renderer.c b/src/v4l2-renderer.c
index 9a17e77..a75a9f9 100644
--- a/src/v4l2-renderer.c
+++ b/src/v4l2-renderer.c
@@ -170,6 +170,14 @@ draw_view(struct weston_view *ev, struct weston_output *output,
 	struct weston_vector p2 = {{ 0.0f, 0.0f, 0.0f, 1.0f }};	// right-bot
 
 	/*
+	 * Check if the surface is still valid. OpenGL/ES apps may destroy
+	 * buffers before they destroy a surface. This check works in the
+	 * serialized world only.
+	 */
+	if (fcntl(vs->planes[0].dmafd, F_GETFD) < 0)
+		return;
+
+	/*
 	 * TODO: We should consider if we make use of 'damage' region.
 	 * In order to do partial update with 'damage', we should
 	 * compose the region with the previous output.
-- 
1.7.9.5

